[alpha.webkit.UncountedLocalVarsChecker] Detect assignments to uncounted local variable and parameters.

rniwa opened the pull request:
This PR updates alpha.webkit.UncountedLocalVarsChecker to emit warnings for assignments to uncounted local variable and parameters instead of just the initialization during the declaration.

Commit:
[alpha.webkit.UncountedLocalVarsChecker] Detect assignments to uncounted local variable and parameters.

This PR updates alpha.webkit.UncountedLocalVarsChecker to emit warnings for assignments to uncounted
local variable and parameters instead of just the initialization during the declaration.

rniwa commented on clang/lib/StaticAnalyzer/Checkers/WebKit/UncountedLocalVarsChecker.cpp:
```diff
@@ -135,7 +135,19 @@ class UncountedLocalVarsChecker
       bool VisitVarDecl(VarDecl *V) {
-        Checker->visitVarDecl(V);
+        auto *Init = V->getInit();
+        if (Init && V->isLocalVarDecl())
```
Oh, interesting. I *think* we do want to check static local variables as well since there is nothing preventing static local variable to be a raw pointer to a ref counted type, which is also bad.

rniwa commented on clang/lib/StaticAnalyzer/Checkers/WebKit/UncountedLocalVarsChecker.cpp:
```diff
@@ -135,7 +135,19 @@ class UncountedLocalVarsChecker
+      bool VisitBinaryOperator(const BinaryOperator *BO) {
+        if (BO->isAssignmentOp()) {
+          if (auto *VarRef = dyn_cast<DeclRefExpr>(BO->getLHS())) {
+            if (auto *V = dyn_cast<VarDecl>(VarRef->getDecl()))
```
Hm... I don't think so. I think any assignment to a global variable is also bad if it's a raw pointer to a ref counted type. It's generally an anti-pattern to avoid.

This makes me think that the checker should probably be renamed to "UncountedVarDeclChecker" or something.

haoNoQ reviewed (approved):
Aha in this case LGTM!

haoNoQ commented:
(It might be a good idea to add comments to those parts of the code to make sure the reader knows that it was intentional.)

rniwa commented:
> (It might be a good idea to add comments to those parts of the code to make sure the reader knows that it was intentional.)

Ok, I was gonna say I'd add tests and then realized that the error message says "local variable" to both static local as well as global variables. That's probably misleading so will correct that.

Commit:
Add tests for assigning to a static local variable and a global variable,
and correct the warning to refer to those variables appropriately instead of
erroneously calling all of them as local variable.
